source/common/upstream/subset_lb.h

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <string>
+
+#include "envoy/common/optional.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/upstream/load_balancer.h"
+
+#include "common/common/macros.h"
+#include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
+#include "common/upstream/upstream_impl.h"
+
+namespace Envoy {
+namespace Upstream {
+
+class SubsetLoadBalancer : public LoadBalancer, Logger::Loggable<Logger::Id::upstream> {
+public:
+  SubsetLoadBalancer(LoadBalancerType lb_type, HostSet& host_set, const HostSet* local_host_set,
+                     ClusterStats& stats, Runtime::Loader& runtime,
+                     Runtime::RandomGenerator& random, const LoadBalancerSubsetInfo& subsets);
+
+  // Upstream::LoadBalancer
+  HostConstSharedPtr chooseHost(LoadBalancerContext* context) override;
+
+private:
+  typedef std::function<bool(const Host&)> HostPredicate;
+
+  // Represents a subset of an original HostSet.
+  class HostSubsetImpl : public HostSetImpl {
+  public:
+    HostSubsetImpl(const HostSet& original_host_set)
+        : HostSetImpl(), original_host_set_(original_host_set), empty_(true) {}
+
+    void update(const std::vector<HostSharedPtr>& hosts_added,
+                const std::vector<HostSharedPtr>& hosts_removed, HostPredicate predicate);
+
+    void triggerCallbacks() { HostSetImpl::runUpdateCallbacks({}, {}); }
+
+    bool empty() { return empty_; };
+
+  private:
+    const HostSet& original_host_set_;
+    bool empty_;
+  };
+
+  typedef std::shared_ptr<HostSubsetImpl> HostSubsetImplPtr;
+
+  typedef std::vector<std::pair<std::string, ProtobufWkt::Value>> SubsetMetadata;
+
+  class LbSubsetEntry;
+  typedef std::shared_ptr<LbSubsetEntry> LbSubsetEntryPtr;
+  typedef std::unordered_map<HashedValue, LbSubsetEntryPtr> ValueSubsetMap;
+  typedef std::unordered_map<std::string, ValueSubsetMap> LbSubsetMap;
+
+  // Entry in the subset hierarchy.
+  class LbSubsetEntry {
+  public:
+    LbSubsetEntry() {}
+
+    LbSubsetMap children_;
+
+    // Only initialized if a match exists at this level.
+    HostSubsetImplPtr host_subset_;
+    LoadBalancerPtr lb_;
+
+    bool initialized() const { return lb_ != nullptr && host_subset_ != nullptr; }
+    bool active() const { return initialized() && !host_subset_->empty(); }
+
+    void initLoadBalancer(const SubsetLoadBalancer& subset_lb, HostPredicate predicate);
+  };
+
+  const LoadBalancerType lb_type_;
```

Comment:
local vars go last (after method definitions). (Apologize if I mistyped this. I can't remember).

Reply:
You did. I'm just having one of those days.